### ISSUE TITLE
Deeplinks plugin obliterates App.entitlements file

### ISSUE THREAD

**Issue by lincolnthree** [ionic-team/ionic-plugin-deeplinks#82]
When using the Deeplinks plugin. It seems to obliterate any other entitlements that might be set in the application. E.g. Push notifications won't work when the .entitlements file this plugin generates is in use. I had to manually add the push-notifications "aps-environment" key to the file in order to get push to work after installing this plugin. I could not enable the entitlement even when toggling the Capability in XCode (the toggle moves, and it says it's enabled, but it is not enabled and nothing happens. Push is still broken):

Here is what my file needed to look like. I understand that this isn't just an issue with this plugin, but I'm not really sure how best to solve this issue so that I don't have to manually edit this file every time I run a build. (I suppose a post-build hook could work, but obviously that's not ideal either.)

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
  <dict>
    <key>com.apple.developer.associated-domains</key>
    <array>
      <string>applinks:www.example.com</string>
    </array>
    <key>aps-environment</key>
    <array></array>
  </dict>
</plist>
```

Actually, on further inspection. It's not just the push entitlements that it obliterates. It's all other entitlements that require configuration in the App.entitlements file.

**Comment by lincolnthree** [comment]
The permissions of this file also seem suspect:

`-rw-r--r--   1 lb3  staff    371 Feb 26 11:18 App.entitlements`

**Comment by lincolnthree** [comment]
Running ionic state reset cleared things up so that I can actually configure the project manually in XCode. I did not re-run another build.

**Comment by lincolnthree** [comment]
Confirmed. Just re-ran the build and this did indeed generate the file that breaks the entitlements. Note that in recent versions of xCode (8+) it appears that entitlements are generally configured entirely via the xcode UI and do not require the xcode entitlements file. It appears if this file exists, the entitlements set in the UI will not be accepted.

Is there a recommended way of setting entitlements that works with this plugin/cordova in general?

**Comment by lincolnthree** [comment]
In fact, I'd love a way to disable generation of this file entirely (without breaking the Android build.)

**Comment by mlynch** [comment]
Any other opinions? I don't feel strongly one way or the other so would like to find a balance here

**Comment by lincolnthree** [comment]
I know you already have my opinion, but I think a flag to disable creation of this file might be a good compromise. People can disable it if they want. Simple, relatively. Not a lot of fuss. Just my 2 cents.

**Comment by FdezRomero** [comment]
**TL;DR:** I love you guys but this broke push on our app silently, and must be fixed.

Every time we uploaded a new IPA to iTunes Connect we got automatic email with a warning saying that we had push notifications active but it was missing from the entitlements. We couldn't find the culprit so we assumed it was a false positive (the capability was on and entitlements set to automatic) and shipped the app. Then we started seeing errors logged in Sentry about devices not registering and found this issue.

Our project currently has 2 entitlement files, one at `/AppName.entitlements` and  another at `/Resources/AppName.entitlements`. I can't remember if we had both initially, but this plugin seems to set the latter as the default from the output in the CLI.

Anyway, I followed @lincolnthree's advice and updated both files to have an `aps-environment` set to `<array/>`. We're no longer getting warnings from iTunes Connect.

This needs to be changed so it doesn't break apps using capabilities. In my opinion, the best way would be to get the existing entitlements file and just add the app-links string to `com.apple.developer.associated-domains` if it's not already present (_It Just Works™_). This may need parsing and modifying XML. Another option would be to just check if the app-link is missing from the entitlements (e.g. via RegExp) and throw an error/warning in the console with a link to the docs explaining how to add it.

**Comment by mlynch** [comment]
I'm thinking of just killing the hooks, they are causing more trouble than they're worth. This is a documentation problem IMO

**Comment by FdezRomero** [comment]
@mlynch I think you're right :+1:

**Comment by fephil** [comment]
For what it's worth, I strongly agree with the above. Removing the entitlement hook would be beneficial as I'd rather use Xcode 8 to set entitlements. I'm also having issues with push notifications.

**Comment by jorrit-wehelp** [comment]
Also, (since Cordova ios 4.3.0 i believe), it's possible to configure the entitlements through plugin.xml/config.xml.
This is done with the Entitlements-Debug.plist and Entitlements-Release.plist files, which are set in the xcode project config.

But this plugin 'clobbers' the xcode project preferences and changes the file it points to. So any plugin that also wants to add/change entitlements breaks.

This is basically the same issue as in the universal-links plugin: https://github.com/nordnet/cordova-universal-links-plugin/issues/89

**Comment by mlynch** [comment]
I killed the hooks. You'll need to manage entitlements yourself in xcode.

**Comment by stalniy** [comment]
@mlynch what about CI builds? Now I need to write such hooks myself....

**Comment by lincolnthree** [comment]
If it were me, I'd want control of the CI build myself anyway. Letting plugins do random things for you that mess with your delivered entitlements is dangerous. Always a good idea to verify the entitlements are exactly as you want them before it ships.

**Comment by mlynch** [comment]
I know this doesn't help anyone right now, but we're working on a new alternative to Cordova that has deeplinking support built in. Part of the approach with Capacitor is that you control your native project, meaning you set the entitlements manually, no more need for scripts to modify it like this, which is incredibly error prone and hard to get right.

I hope, when it's ready (early 2018) that it'll be helpful. 

https://github.com/ionic-team/capacitor

**Comment by stalniy** [comment]
There is a way to do this via `edit-config` directive in `config.xml`

```xml

        <edit-config file="*-Release.plist" mode="merge" target="com.apple.developer.associated-domains">
            <array>
                <string>applinks:misto.freaksidea.com</string>
            </array>
        </edit-config>
```

and related cordova issue https://issues.apache.org/jira/browse/CB-13564

**Comment by lincolnthree** [comment]
@mlynch I've noticed the Capacitor Repo, and am intrigued. Is this something like NativeScript? Replacing Cordova seems like a ... very lofty goal.

**Comment by mlynch** [comment]
@lincolnthree we're all about lofty goals here 😅 

It's inspired by popular tools in this space, including Cordova, React Native, and Turbolinks. It's probably least like NativeScript since their API approach is probably least portable to the web..

There's a lot of details that we'll share in time, but the goal is to make it easier to build web-first apps that run just as well on iOS and Android as they do as Progressive Web Apps. Compared to Cordova, there will be more plugins out of the box that will be ready to use, and we will have interopt with existing Cordova plugins.

Additionally, we will be offering a set of supported, premium plugins for businesses that would like to pay for more assurances around support and platform updates, something that is challenging when relying on the 3rd party plugin ecosystem.

Basically, we are trying to make the native layer experience better and easier to use, and also easier to build and find maintained native plugins.

**Comment by lincolnthree** [comment]
@mlynch That's super interesting. I hope it's fairly easy to migrate - With a large app that sounds (without more information) pretty daunting!

**Comment by mlynch** [comment]
@lincolnthree we have a ways before you'll need to make that decision. I'd like to have a middle ground where you're able to use our plugins in your Cordova app, as a transition to the tooling and such. We need to get this in the wild and testing w/ early adopters willing to use it on new projects to learn more before we even consider something like that.

If you'd like to stay in the loop, we have a mailing list where we'll be posting updates: https://ionic-team.github.io/capacitor/

🍻

**Comment by chrisweight** [comment]
This has just been published as part of https://github.com/ionic-team/ionic-plugin-deeplinks/tree/v1.0.16
